### PR TITLE
STORM-1369: Add MapState implementation to storm-cassandra.

### PR DESCRIPTION
**Pull request by mkoch1**
STORM-1369: Add MapState implementation to storm-cassandra.

https://issues.apache.org/jira/browse/STORM-1369

### CODE REVIEW

File: external/storm-cassandra/src/main/java/org/apache/storm/cassandra/executor/AsyncExecutor.java [hunk 1]
```diff
@@ -23,16 +23,16 @@
...
+import org.apache.storm.topology.FailedException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.Serializable;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
+import java.util.*;
```
Comment (revans2) [inline external/storm-cassandra/src/main/java/org/apache/storm/cassandra/executor/AsyncExecutor.java]: In general we prefer to not use the wildcard on includes.

File: external/storm-cassandra/src/main/java/org/apache/storm/cassandra/executor/AsyncExecutor.java [hunk 2]
```diff
@@ -138,6 +139,109 @@ public void onFailure(Throwable t) {
...
+
+        final SettableFuture<List<T>> settableFuture = SettableFuture.create();
+        final AsyncContext<T> asyncContext = new AsyncContext<>(inputs, maxParallelRequests, settableFuture);
+
+        for (int i = 0; i < statements.size(); i++) {
+
+            // Acquire a slot
+            if (asyncContext.acquire()) {
+
+
```
Comment (revans2) [inline external/storm-cassandra/src/main/java/org/apache/storm/cassandra/executor/AsyncExecutor.java]: nit: don't need the extra line

File: external/storm-cassandra/src/main/java/org/apache/storm/cassandra/executor/AsyncExecutor.java [hunk 3]
```diff
@@ -138,6 +139,109 @@ public void onFailure(Throwable t) {
...
+                            }
+                        }
+
+                        @Override
+                        public void onFailure(Throwable throwable) {
+                            handler.failure(throwable, input);
+                            asyncContext
+                                    .exception(throwable)
+                                    .release();
+                            LOG.error(String.format("Failed to execute statement '%s' ", statement), throwable);
```
Comment (revans2) [inline external/storm-cassandra/src/main/java/org/apache/storm/cassandra/executor/AsyncExecutor.java]: do we need a `pending.decrementAndGet();` here too?

File: external/storm-cassandra/src/main/java/org/apache/storm/cassandra/trident/state/CassandraBackingMap.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        this.conf = conf;
+        this.options = options;
+        List<String> allFields = options.keyFields.toList();
+        allFields.addAll(options.stateMapper.getStateFields().toList());
+        this.allFields = new Fields(allFields);
+    }
+
+    @Override
+    public List<T> multiGet(List<List<Object>> keys) {
+        LOG.info("multiGet fetching {} values.", keys.size());
```
Comment (revans2) [inline external/storm-cassandra/src/main/java/org/apache/storm/cassandra/trident/state/CassandraBackingMap.java]: This should probably be debug.

File: external/storm-cassandra/src/main/java/org/apache/storm/cassandra/trident/state/CassandraBackingMap.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            states.add(state);
+        }
+
+        return states;
+
+    }
+
+    @Override
+    public void multiPut(List<List<Object>> keys, List<T> values) {
+        LOG.info("multiPut writing {} values.", keys.size());
```
Comment (revans2) [inline external/storm-cassandra/src/main/java/org/apache/storm/cassandra/trident/state/CassandraBackingMap.java]: debug

File: external/storm-cassandra/src/main/java/org/apache/storm/cassandra/executor/AsyncExecutor.java (hunk 3 above)
Comment (mkoch1) [inline external/storm-cassandra/src/main/java/org/apache/storm/cassandra/executor/AsyncExecutor.java]: Good catch, it should also handle exceptions like the onSuccess method.

**Comment by mkoch1** [comment]
Pls note that mutliPut needs an update - will push commit today or tomorrow.

**Comment by mkoch1** [comment]
That should do it - the batching option has been removed from MapState, in favor of parallel processing with existing opaque and transactional logic to handle consistency. Cassandra batch statements are more trouble than they're worth.

**Comment by ptgoetz** [comment]
Unfortunately this patch uses Cassandra-unit, which is LGPL licensed, so in it's current state we can't accept it. For that reason I am -1. We've run into this before, and I've lobbied the Cassandra-unit author to switch to an Apache compatible license (https://github.com/jsevellec/cassandra-unit/issues/163) to no avail.

We can, however, accept it if the Cassandra-unit dependency and tests that depend on it are removed.

Again, it's unfortunate. If more people encourage the Cassandra-unit author to choose a different license, perhaps he'll consider it.

**Comment by mkoch1** [comment]
Did not notice that, sorry. Replaced now with an external resource implementation. The code is not all that complex so it's not a big deal. 

One thing I should point out: I borrowed the base implementation from a different project, also under Apache 2.0 license. I've included their copyright line in the file: https://github.com/apache/storm/pull/1781/commits/e46496337af84a610ef742c8cf5988d8dab19e47#diff-a8ce0c2f5e3ce0193d3ebc49f54301fa

Original code: https://github.com/spring-projects/spring-insight-plugins/blob/master/collection-plugins/cassandra12/src/test/java/com/springsource/insight/plugin/cassandra/embeded/EmbeddedCassandraService.java

Please let me know if there are any issues with this.

**Comment by ptgoetz** [comment]
@mkoch1 Thanks for addressing the license issue. The way you handled the borrowed code looks fine, the important part is to leave the original copyright statement, which you did. Thanks for pointing that out.

+1

**Comment by HeartSaVioR** [comment]
@mkoch1 Could you rebase this? After rebasing I'll merge this.

**Comment by mkoch1** [comment]
Rebase is done, sorry for the delay.

**Comment by HeartSaVioR** [comment]
@mkoch1 
Could you squash commits into one if you don't mind? I know it's more than 2000 lines change but looks like separating commits are not needed.
And I would like to wait for 1.1.0 RC vote to pass to avoid any conflicts with the release. That says, this is not going to be included to 1.1.0.

**Comment by mkoch1** [comment]
Squashed. I don't really see how this should cause any conflicts though. The only changes in existing code are a couple of namespace corrections (to avoid wildcard imports, as suggested earlier) and documentation. Other than that, it's all new code in new classes.

**Comment by HeartSaVioR** [comment]
+1

**Pull request merged**